### PR TITLE
Wire cert contexts

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
#### What type of PR is this?

/kind cleanup
#### What this PR does / why we need it:
All the controllers should use context for signalling termination of communication with API server. Once kcm cancels context all the cert controllers which are started via kcm should cancel the APIServer request in flight instead of hanging around.
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
cc @damemi. Can you check if I missed anything?

**Review by damemi — COMMENTED** [review]
@ravisantoshgudimetla one thing that I missed when doing a lot of these is that the end goal is to wire the new contexts through to all the API calls as well (so replacing every `context.TODO()`). I don't know if @soltysh still wants to do that now, or hold off for the next release. 

But just grepping through your branch I found the following that could be relevant, so you may want to check these out to make sure they trace back to the new contexts:
```
~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" pkg/controller/certificates/*
pkg/controller/certificates/approver/sarapprove.go:135:	sar, err := a.client.AuthorizationV1().SubjectAccessReviews().Create(context.TODO(), sar, metav1.CreateOptions{})
pkg/controller/certificates/certificate_controller_test.go:51:		_, err := client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
pkg/controller/certificates/cleaner/cleaner.go:108:		if err := ccc.csrClient.Delete(context.TODO(), csr.Name, metav1.DeleteOptions{}); err != nil {
pkg/controller/certificates/rootcacertpublisher/publisher.go:192:		_, err = c.client.CoreV1().ConfigMaps(ns).Create(context.TODO(), &v1.ConfigMap{
pkg/controller/certificates/rootcacertpublisher/publisher.go:227:	_, err = c.client.CoreV1().ConfigMaps(ns).Update(context.TODO(), cm, metav1.UpdateOptions{})
pkg/controller/certificates/signer/signer.go:170:		_, err = s.client.CertificatesV1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})
pkg/controller/certificates/signer/signer.go:184:	_, err = s.client.CertificatesV1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})

~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/*
staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller.go:187:	configMap, err := c.client.CoreV1().ConfigMaps(c.configmapNamespace).Get(context.TODO(), c.configmapName, metav1.GetOptions{})

~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/server/options/*
staging/src/k8s.io/apiserver/pkg/server/options/serving.go:360:	ln, err := config.Listen(context.TODO(), network, addr)

~/go/src/k8s.io/kubernetes$ grep -rni "context\.TODO" staging/src/k8s.io/kube-aggregator/pkg/apiserver/*
staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:307:					context.TODO(), s.GenericAPIServer.APIServerID, metav1.GetOptions{})
```

(note there could be more, but I just searched through the folders you've already edited)

**Comment by ravisantoshgudimetla** [comment]
@damemi - Thanks for the review. I was initially thinking the same. However as you might have already noticed that - if the context is cancelled, Run anyway is going to stop, so rest of the methods even if they're using the context.TODO() may not have effect. To me passing `ctx` to New methods did not make much sense. I wish the controller's handlers are decoupled from `CertificateController` [struct](https://github.com/kubernetes/kubernetes/blob/704dc9a99e6a499a8b36d4637b99b63960e36837/pkg/controller/certificates/certificate_controller.go#L53-L52), so that we can call them via `Run` so that we have a single entry point and no initialization. For example, we can make CertificateController an interface and Run method there.

I did not want to make all the changes in a single PR and make it hard to review. I can comeback to this later if you and @soltysh are fine with the approach I suggested.

**Comment by soltysh** [comment]
@ravisantoshgudimetla  `Run` won't be cancelled if it hangs on a client call, that's why we need to pass the context all the way down to all client invocation. I'm looking at the PR and I'm wondering if we could split it into smaller bits, like just the apiserver bits and necessary places using `DynaminFileCAContent` and split controllers along with the changes requested by @damemi in a followup?

File: cmd/kubelet/app/auth.go
Comment (aojea) [inline cmd/kubelet/app/auth.go:107]: I'm not able to understand well this pattern, does this means that the context can be cancelled from the inside?

Comment (ravisantoshgudimetla) [reply cmd/kubelet/app/auth.go:107]: Yeah, the goal is to have a routine which looks at the stopChannel and cancels the context when stopChannel is closed.

**Comment by ravisantoshgudimetla** [comment]
@damemi Not sure if you looked at the right branch but most of the changes should be already in place. I was suggesting how it be better in the long run.

```
➜  kubernetes git:(wire-contexts-certificates) grep -rni "context\.TODO" pkg/controller/certificates/*
pkg/controller/certificates/approver/sarapprove_test.go:134:			ctx := context.TODO()
pkg/controller/certificates/cleaner/cleaner_test.go:230:			ctx := context.TODO()
pkg/controller/certificates/signer/signer_test.go:296:			ctx := context.TODO()
➜  kubernetes git:(wire-contexts-certificates) grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/*
staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:224:			ctx := context.TODO()
```

File: cmd/kubelet/app/auth.go
Comment (tkashem) [reply cmd/kubelet/app/auth.go:107]: > yeah, but AFAIK the context is meant to be cancelled from the parent not from the childs,

yes, this is the recommended use, but here the lifecycle aspect of the context is scoped within the go routine, it's because only when the `stopCh` is closed should we cancel the context. If we return the cancel func to the caller this is no longer guaranteed.

I would recommend:
```
go func() {
                        defer cancel()
			select {
			case <-stopCh:
			case <-ctx.Done():
			}
           }
```

Comment (tkashem) [reply cmd/kubelet/app/auth.go:107]: i take it back, we don't need a defer in this case.

**Comment by ravisantoshgudimetla** [comment]
Made some changes after picking Mike's commits:

```
➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" pkg/controller/certificates/*
pkg/controller/certificates/approver/sarapprove_test.go:134:			ctx := context.TODO()
pkg/controller/certificates/certificate_controller_test.go:50:		_, err := client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
pkg/controller/certificates/certificate_controller_test.go:72:	ctx := context.TODO()
pkg/controller/certificates/cleaner/cleaner_test.go:229:			ctx := context.TODO()
pkg/controller/certificates/rootcacertpublisher/publisher_test.go:158:			ctx := context.TODO()
pkg/controller/certificates/rootcacertpublisher/publisher_test.go:267:			ctx := context.TODO()
pkg/controller/certificates/signer/signer_test.go:298:			ctx := context.TODO()

========================================================================


➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/*
staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:225:			ctx := context.TODO()
=========================================================================


➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" staging/src/k8s.io/apiserver/pkg/server/options/*
staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:393:	ctx := context.TODO()
staging/src/k8s.io/apiserver/pkg/server/options/serving.go:360:	ln, err := config.Listen(context.TODO(), network, addr)
➜  kubernetes git:(wire-cert-contexts) ✗ grep -rni "context\.TODO" staging/src/k8s.io/kube-aggregator/pkg/apiserver/*
staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:249:		ctx := context.TODO()
staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:320:					context.TODO(), s.GenericAPIServer.APIServerID, metav1.GetOptions{})

We need to pass a dummy context but the controller is not making any api calls.
```

I don't mind splitting the PR's commits into 2 different commits. Let me know if  it makes reviews easier.

File: cmd/kubelet/app/auth.go
Comment (aojea) [reply cmd/kubelet/app/auth.go:107]: you are only blocking on the stopCh to cancel the context
```
		ctx, cancel := context.WithCancel(context.Background())
		go func() {
			<-stopCh
			cancel() // stopCh closed, so cancel our context
			}
```
the only way this goes out is if someone cancels the context, but that will mind that it has to be done from within the children

File: staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:226]: nit: `target.RunOnce(context.TODO())`

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:394]: nit:  `dynamicRequestHeaderProvider.RunOnce(context.TODO())`

File: pkg/controlplane/instance.go
Comment (tkashem) [inline pkg/controlplane/instance.go:458]: nit: is there any common place where we can define this function and use it here?

**Review by tkashem — COMMENTED** [review]
LGTM (just have a few minor nits)

**Review by soltysh — APPROVED** [review]
/lgtm
/approve
/release-note none
/kind feature

**Comment by rphillips** [comment]
PR description needs more context... Looks like a cleanup PR?

**Comment by rphillips** [comment]
lgtm after that update.

**Review by mrunalp — APPROVED** [review]

**Pull request merged**
